Dev datamigration microsoft.data migration 2020 09 01 preview

vasundhraManhas opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that label “WaitForARMFeedback” will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits.

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

vasundhraManhas committed:
Adds base for updating Microsoft.DataMigration from version stable/2021-06-30 to version 2020-09-01-preview

vasundhraManhas committed:
Updates readme

vasundhraManhas committed:
Updates API version in new specs and examples

Commit:
add examples and change swagger

Commit:
camel casing error (removing the resource since it is redundant)

Commit:
added "systemData" property to the response of sqlMigrationService

Commit:
adding isDataAction property in operationsDefinition

ashutoshsuman99 commented:
Adding @rchlkm

Commit:
fixing Avocado error : removing unreferenced files

Commit:
styling changes

Commit:
adding systemData to DatabaseMigrations

Commit:
suppressing lintDiff

hitenjava commented on specification/azure-kusto/resource-manager/Microsoft.Kusto/stable/2020-06-14/examples/KustoAttachedDatabaseConfigurationsCreateOrUpdate.json:
```diff
@@ -1,61 +1,61 @@
-{
-  "parameters": {
-    "subscriptionId": "12345678-1234-1234-1234-123456789098",
```
Looks like some git merge issue or formatting change?
Why is this file showing up - This should not be changed.

vasundhraManhas replied:
There was some styling/formatting error (Prettier check) and the suggestion  was to use npm run prettier-fix
I think it fixed the formatting issues in all the files

Commit:
Revert "styling changes"

This reverts commit 4cba4308e6b3b4f35144f6ae28b4c760cd6f90f8.

Commit:
formatting changes (after running npm run prettier)

Commit:
adding v1 apis

Commit:
adding v1 apis

Commit:
Revert "adding v1 apis". Bad update

This reverts commit 99b1f84465749a4948ece4cacfb63652b0cd5a4e.

Commit:
Revert "adding v1 apis". Bad update

This reverts commit 795344a819d2d72f3c86c3714274df48f5303205.

Commit:
adding v1 paths in v2 spec

Commit:
adding stable-2018 apis

Commit:
updating api version in definitions

Commit:
avocado error :  removing unreferenced files

Commit:
suppressing lintdiff errors R4013 and R4037

Commit:
removing subscriptioid parameter from v1 and
referencing to susbscriptionid parameter from v2 (to resolve conflict)

Commit:
adding v1 apis as a separate json file

Commit:
adding a missing example for v1 spec

Commit:
removing duplicate api (get operation) from v1 spec

Commit:
Revert "removing duplicate api (get operation) from v1 spec"

This reverts commit 545e6ff29239d3b33f393af8b8c4fd4021cebe1f.

Commit:
avocado error : api versions correction

Commit:
avocado error : adding definitions in readme and changing api version in v2 spec to 2020-09-01-preview

Commit:
adding missing definitions and removing extra definitions in readme

Commit:
removing operation path from v1 spec

j5lim commented on specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2020-09-01-preview/examples/SqlMiCreateOrUpdateDatabaseMigrationMIN.json:
```diff
@@ -0,0 +1,82 @@
+      "location": "northeurope",
+      "properties": {
+        "sourceDatabaseName": "aaa",
+        "scope": "/subscriptions/00000000-1111-2222-3333-444444444444/resourceGroups/testrg/providers/Microsoft.Sql/managedInstances/instance",
```
>scope [](http://example.com/codeflow?start=9&length=5)

Is this different from the scope parameter?

This should be returned in the response. (all properties in the put should be returned except secrets)

j5lim replied:
Why scope needs to be specified both in the URL (line 3) and the body (line 10)? Are they different?

If the scope in the body is needed in the request body, the response also should return it.

j5lim commented on specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2020-09-01-preview/sqlmigration.json:
```diff
@@ -0,0 +1,1718 @@
+          }
+        }
+      },
+      "put": {
```
>"put": [](http://example.com/codeflow?start=6&length=6)

No PATCH? If a partial update is supported, it must be via PATCH.

[https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc009-use-put-for-update-of-entire-resource-ie-replace-use-patch-only-for-partial-update-of-resource](https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc009-use-put-for-update-of-entire-resource-ie-replace-use-patch-only-for-partial-update-of-resource)

ashutoshsuman99 replied:
Currently we don't support partial update on Database Migration resource. @hitenjava can you please add further details on this.

hitenjava replied:
Correct - databaseMigrations resource is a proxy resource and we don't support partial update on databaseMigrations resource.

Commit:
trial : trying merging with 2021 stable version

Commit:
nit errors in readme

Commit:
updating api versions in definitions of 2021 stable version

Commit:
removing operations api from v1 (stable 2021)

Commit:
changing sqlMigrationInfo to sqlMigrationInformation in v2 swagger

Commit:
Renaming FileShare to SqlFileShare in v2 spec

Commit:
Renamimg BackupFileInfo to SqlBackupFileInfo in v2 spec

Commit:
renaming BackupSeiInfdo to SqlBackupSetInfo in v2 spec

Commit:
Renaming TrackedResource to TrackedResourceARM in the v1 spec

Commit:
deleting duplicate subscriptionId parameter in v1

Commit:
sqlmigration casing error

Commit:
referencing SystemData definition from the v2 spec

Commit:
avacado error : removing unreferenced file Operations_List.json

Commit:
adding custom-words.txt file

Commit:
reverting adding custom-words.txt

Commit:
exchanging positions of sqlmigration and datamigration in readme

Commit:
adding sqlmigration in custom-words.txt

Commit:
adding systemData to TrackedResource in v2
removing TrackedResourceARM from v1 and referring to v2 TrackedResource instead

Commit:
adding "x-ms-azure-resource": true extension in DataMigrationService and Project definitions in v1

Commit:
changing veersion from 2020-09-01-preview to "2021-10-30-preview"

Commit:
manually changing api version (to 2021-10-30)in readme

jianyexi commented:
/azp run

Commit:
adding ooperations api to v1 spec

Commit:
adding the subscriptionId parameter in v1 spec

Commit:
adding operations_list in the examples (for v1)

Commit:
Revert "adding ooperations api to v1 spec", adding the subscriptionId parameter in v1 spec, adding operations_list in the examples (for v1)

This reverts commit b94a96d8f48edef45967a57050e09eb27f9ea0de, 40e80e0d2a8e9fab72443e3893a9f0712c224f0a, 7be34463aaf58418f4574d3df9288e64cd27c5e4

Commit:
removing x-ms-azure-resource extension

Commit:
Revert "removing x-ms-azure-resource extension"

This reverts commit 8bf42972d7a1f3e5f8f1f4bf0d1fad19b77a161c.

Commit:
add subscriptionId to v1 spec

Commit:
Revert "add subscriptionId to v1 spec"

This reverts commit e3e729be0b1124875e279d9d718d8226dba2f8b8.

Commit:
removing read only : true for systemData properties

Commit:
setting readonly to true in systemData definition in v2 swagger.
(note : the chnages in v2 swagger are done manually for now, change these from the RP side too)

Commit:
renaming the specs to check if having the name datamigration.json has some importance

Commit:
Revert "renaming the specs to check if having the name datamigration.json has some importance"

This reverts commit a68c0c07d2a3f9932310eb910a683d1e824d7dfc.

Commit:
adding "x-ms-azure-resource": true to trackedResource and removing it from
datamigrationService and Project definition as they extend from trackedResource

Commit:
correctly placing "x-ms-azure-resource": true

Commit:
prettier check error on sqlmigration.json

Commit:
referencing systemData from common-types/resource-management/v2/types.json#/definitions
instead from v2 spec (to see if that affects the systemData error in cross-version)

Commit:
Revert "referencing systemData from common-types/resource-management/v2/types.json#/definitions"

This reverts commit 94641f1b36cc7828ea8efd21f4df89c282f51941.

Commit:
keeping subscriptionId parameter in both specs (by renaming it in v2)

Commit:
renaming subscriptionId to subscriptionIdParameter in the paths of v2

Commit:
casing error

Commit:
using subscriptionIdParameter instead of subscriptionId in the examples of v2

Commit:
renaming subscriptionId to subscriptionidParameter in v2 example

Commit:
Reverting the changes related to keeping subscriptionId parameter in both specs

This reverts commit fa22043a2b4ca284466bacda244f19c087a78718,
13d549928711b2d85e776f0680a7347242582ac2,
2755f643ed32b766fbb98ed6a7fd814211aecea3,
6e74171367ddf0a4a9f7a6b36b8f41a65a2c7701,
f3142b41b841793ae409a7e2d89d9bc4c262e9e7

vasundhraManhas commented:
/azp run

vasundhraManhas committed:
update tag in readme

vasundhraManhas commented:
/azp run

ArcturusZhang commented:
/azp run

Commit:
renaming trackedResource to trackedResourceARM in v2 and restoring the trackedResource in v1

Commit:
Merge branch 'dev-datamigration-Microsoft.DataMigration-2020-09-01-preview' of https://github.com/vasundhraManhas/azure-rest-api-specs into dev-datamigration-Microsoft.DataMigration-2020-09-01-preview

ArcturusZhang commented:
/azp run

vasundhraManhas commented:
/azp run

Commit:
Revert "renaming trackedResource to trackedResourceARM in v2 and restoring the trackedResource in v1"

This reverts commit 269a4c3305c6e282bfa192b162b959d1b8083033.

Commit:
removing system data and x-ms-azure extension from tracked resource (because making changes in the tracked resource will involove dsMianDev repo changes)

Commit:
adding "x-ms-azure-resource": true to trackedResource of v1

Commit:
adding "x-ms-azure-resource": true to dataMigration service and project
(because of the error : The 200 response model for an ARM PUT operation must have x-ms-azure-resource extension set to true in its hierarchy. Operation: 'Services_CreateOrUpdate' Model: 'DataMigrationService'.
Location: Microsoft.DataMigration/preview/2021-10-30-preview/datamigration.json#L97)

also, removing x-ms-azure-resource from trackedResource

Commit:
Revert "adding "x-ms-azure-resource": true to trackedResource of v1"

This reverts commit bd0fb9c3bb96a736712baaa71efd2e91c49375c0.
and f2440eef137d7771217b9924ce941e876363afe9
(bad commit)

vasundhraManhas commented:
/azp run

Commit:
Revert "removing system data and x-ms-azure extension from tracked resource (because making changes in the tracked resource will involove dsMianDev repo changes)"

This reverts commit e330db6a6abe2209d8d28a07da295267a036a9de.
SystemData is required to be in trackedResource

Commit:
Adding Scope Parameter to examples

Commit:
Revert "Adding Scope Parameter to examples"

This reverts commit b6c827585bb509306f2082e5686b31ee0333b182.

Commit:
Adding Scope parameter in response of SqlMiGetDatabaseMigration.json

Commit:
Adding Scope Parameter to other examples

Commit:
ARM:
Adding x-ms-secret: true to password (SqlFileshare and SqlConnectionInfomation), accountkey, authkey1 and authkey2

Commit:
Adding x-ms-pageable to DatabaseMigrations_ListByScope, Operations_List, SqlMigrationServices_ListByResourceGroup, SqlMigrationServices_listMigrations, SqlMigrationServices_ListBySubscription

Commit:
Merge branch 'AddingX-MS-SECRET' into dev-datamigration-Microsoft.DataMigration-2020-09-01-preview

Commit:
Removing x-ms-secret from Authkeys

Commit:
Adding x-ms-secret to authKeys in AuthenticationKeys and RegenAuthKeys defination

Commit:
Suppressing Secret_Property model validation error

Commit:
Adding suppress SECRET_PROPERTY in Suppression section

Commit:
changing GET getMonitoringData to POST listMonitoringData

Commit:
Merge branch 'dev-datamigration-Microsoft.DataMigration-2020-09-01-preview' of https://github.com/vasundhraManhas/azure-rest-api-specs into dev-datamigration-Microsoft.DataMigration-2020-09-01-preview

vasundhraManhas commented on specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2021-10-30-preview/sqlmigration.json:
```diff
@@ -0,0 +1,1718 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataMigration/sqlMigrationServices/{sqlMigrationServiceName}/getMonitoringData": {
```
@pilor GET getMonitoringData call has been changed to POST "listMonitoringData" call

Commit:
Removing location property from DatabaseMigration definition

ashutoshsuman99 commented on specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2021-10-30-preview/sqlmigration.json:
```diff
@@ -0,0 +1,1718 @@
+        }
+      ],
+      "properties": {
+        "location": {
```
Hi @pilor, we have removed the location property for DatabaseMigration.

Commit:
New Cancel\Cutover operations with MigrationOperationId in Request body.

Commit:
Merge branch 'dev-datamigration-Microsoft.DataMigration-2020-09-01-preview' of https://github.com/vasundhraManhas/azure-rest-api-specs into dev-datamigration-Microsoft.DataMigration-2020-09-01-preview

Commit:
updating examples for cancel and cutover operations

vasundhraManhas commented:
azp run

vasundhraManhas commented:
/azp run

Commit:
Chaning SqlMigrationService_GetResourceState to SqlMigrationService_Get

Commit:
Revert "Chaning SqlMigrationService_GetResourceState to SqlMigrationService_Get"

This reverts commit d16dc91a39245cecc4289b206af070eccb74b783.

Commit:
Changing GetResourceState to Get

Commit:
Adding 2 namespaces 2 SqlMi Migration API and updating examples

Commit:
Add Vm APIs, change operation Id of Mi and VM APIs and Update examples

Commit:
Suppressing list by subscription DatabaseMigration LintDiff error

Commit:
Adding discriminator to databaseMigrationProperties in order to separate SqlMi and SqlVm migrations

Commit:
[nit error] changing SqlConnectionInfo to SqlConnectionInformation

Commit:
nit error

Commit:
updating the example SqlMiGetDatabaseMigration.json to use"kind" in order to resolve the error : Additional properties not allowed

Commit:
updating all Mi and Vm examples to use "kind" in the properties

Commit:
styling changes (for prettier check)

Commit:
updating listMigrationsByService.json xample to use kind property

Commit:
running prettier check

Commit:
adding other properties in request along with kind in sqlVmCreateOrUpdateMax.json example

Commit:
Revert "adding other properties in request along with kind in sqlVmCreateOrUpdateMax.json example"

This reverts commit 8a47679fa45846b8bc41d7651c7c91e2f7ef943c.

Commit:
keeping both kind and other properties in the request body for sqlMiCreateOrUpdateDatabaseMigrationMax.json

Commit:
keeping both kind and other properties in all the examples

Commit:
adding enums for commandType and taskType discriminator

ArcturusZhang commented:
The cross version breaking changes are false alarms

raych1 commented:
@ArcturusZhang , `model validation` has errors in examples. Please fix them.
https://github.com/Azure/azure-rest-api-specs/pull/16032/checks?check_run_id=4439591756

Commit:
adding x-ms-enum for commandType and taskType (v1)

Commit:
adding remaining enum values for taskType and commandType

vasundhraManhas commented:
> @ArcturusZhang , `model validation` has errors in examples. Please fix them. https://github.com/Azure/azure-rest-api-specs/pull/16032/checks?check_run_id=4439591756

You are referring to an earlier commit. These errors have been fixed.

Commit:
Removing List By Scope APIs for Mi and VM

Commit:
Remove Delete APIs for SqlMi and SqlVm

Commit:
Making DatabaseMigrationSqlVm_Cancel a x-ms-long-running-operation

Commit:
Moving common properties to DatabaseMigrationProperties

Commit:
Creating ResourceClass for SqlMi and SqlVm and updating the APIs accordingly

Commit:
Suppressing LintDiff errors for SqlMi and SqlVm Resource Classess

Commit:
Making x-ms-client-flatten: false for DatabaseMigration and other SqlMi and SqlVm Resource Class

Commit:
updating parameters in examples: Services_checkChildrenNameAvailability.json and Services_CheckNameAvailability

Commit:
Merge branch 'dev-datamigration-Microsoft.DataMigration-2020-09-01-preview' of https://github.com/vasundhraManhas/azure-rest-api-specs into dev-datamigration-Microsoft.DataMigration-2020-09-01-preview

Commit:
v1 s360 errors

Commit:
object_missing_required_property error: updating example services_Start.json

Commit:
revert 81733a02cc88d52d69497fff3a5b7abafadedeee

Commit:
prettier check

Commit:
reverting all chnages related to  "v1 s360 errors"

Commit:
v1 s360 errors: adding missing properties to mongoDbConnectionInfo in common.json

Commit:
nit error

Commit:
v1 s360 errors: adding missing properties to sqlConnectionInfo in common.json

Commit:
v1 s360 errors: adding missing properties to mySqlConnectionInfo in common.json

Commit:
v1 s360 errors: adding missing properties to PostgreSqlConnectionInfo in common.json

Commit:
v1 s360 errors: adding missing properties in MigratePostgreSqlAzureDbForPostgreSqlSyncTask.json

Commit:
v1 s360 errors: removing problematic property from MigratePostgreSqlAzureDbForPostgreSqlSyncTask

Commit:
v1 errors: adding missing properties in MigrateSchemaSqlServerSqlDbTask.json

Commit:
v1 s360 errors: adding missing properties in MigrateSqlServerSqlDbTask.json
and MigrateSqlServerSqlMITask.json

Commit:
v1 360 errors: adding missing properties in projects.json

vasundhraManhas commented:
/azp run

Commit:
v1 s360 errors: adding missing properties in Services.json

Commit:
v1 s360 error: adding missing properties in Tasks.json

Commit:
v1 s360 errors: adding missing properties in TasksCommon.json

Commit:
v1 s360 error: adding missing properties to MigratePostgreSqlAzureDbForPostgreSqlSyncTask.json

Commit:
v1 s360 errors: adding items to events in "MigratePostgreSqlAzureDbForPostgreSqlSyncTaskOutputError"

Commit:
v1 s360 errors: removing required property

Commit:
prettier check

ash1625 commented on specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2021-10-30-preview/definitions/Common.json:
```diff
@@ -291,6 +291,22 @@
+          "type": "string",
+          "description": "name of the server"
+        },
+        "serverBrandVersion": {
```
Didn't see serverBrandVersion flagged in s360 report for SqlConnectionInfo object

vasundhraManhas replied:
> Didn't see serverBrandVersion flagged in s360 report for SqlConnectionInfo object

@ash1625 thanks, removed now

Commit:
nit

raych1 commented:
/azp run

msyyc commented:
/azp run
